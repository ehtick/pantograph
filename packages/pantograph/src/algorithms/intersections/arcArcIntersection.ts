import { Vector } from "../../definitions.js";
import { Arc } from "../../models/segments/Arc.js";
import removeDuplicatePoints from "../../utils/removeDuplicatePoints.js";
import {
  add,
  distance,
  normalize,
  perpendicular,
  sameVector,
  scalarMultiply,
  subtract,
} from "../../vectorOperations";

const complementArc = (arc: Arc): Arc => {
  const { firstPoint, lastPoint, center, clockwise } = arc;
  return new Arc(lastPoint, firstPoint, center, clockwise, {
    ignoreChecks: true,
  });
};

const handleOverlaps = (arc1: Arc, arc2: Arc): Arc[] => {
  // handle the case with common points at start or end first
  if (arc1.isSame(arc2)) {
    return [arc1];
  }
  // two arcs with common points at start and end (but not the same arc)

  const points = removeDuplicatePoints(
    [
      arc2.isOnSegment(arc1.firstPoint) ? arc1.firstPoint : null,
      arc2.isOnSegment(arc1.lastPoint) ? arc1.lastPoint : null,
      arc1.isOnSegment(arc2.firstPoint) ? arc2.firstPoint : null,
      arc1.isOnSegment(arc2.lastPoint) ? arc2.lastPoint : null,
    ].filter((p) => p !== null) as Vector[],
    // We sort by the param value of the first arc. This means that the points
    // will be sorted with the same orientation than arc1
  ).sort((a, b) => arc1.pointToParam(a) - arc1.pointToParam(b));

  if (points.length === 0) return [];
  // We consider the case when the arcs touch only on
  // the last point. We consider that they do not overlap there
  //
  // We might want to revisit this choice
  else if (points.length === 1) return [];
  else if (points.length === 2) {
    // Similar to the case with length 1, we ignore the double overlapping point
    if (arc1.isSame(complementArc(arc2))) return [];
    return [new Arc(points[0], points[1], arc1.center, arc1.clockwise)];
  } else if (points.length === 3) {
    // Similar to the case with length 1, we ignore the single overlapping point
    const startIndex =
      sameVector(points[0], arc2.lastPoint) ||
      sameVector(points[0], arc2.firstPoint)
        ? 1
        : 0;
    return [
      new Arc(
        points[0 + startIndex],
        points[1 + startIndex],
        arc1.center,
        arc1.clockwise,
      ),
    ];
  } else if (points.length === 4) {
    return [
      new Arc(points[0], points[1], arc1.center, arc1.clockwise),
      new Arc(points[2], points[3], arc1.center, arc1.clockwise),
    ];
  }
  throw new Error("Bug in the arc arc overlap algorithm");
};

export function arcArcIntersection(
  arc1: Arc,
  arc2: Arc,
  includeOverlaps = false,
  precision?: number,
): Vector[] | Arc[] {
  const epsilon = precision ? precision : arc1.precision;
  const centersDistance = distance(arc1.center, arc2.center);

  const radiusSum = arc1.radius + arc2.radius;

  // The circles do not touch
  if (centersDistance > radiusSum + epsilon) {
    return [];
  }

  const radiusDifference = Math.abs(arc1.radius - arc2.radius);

  // The arcs are concentric
  if (centersDistance < radiusDifference - epsilon) {
    return [];
  }

  // With a common center we can have overlaps
  if (centersDistance < epsilon) {
    if (radiusDifference > epsilon) {
      return [];
    } else {
      if (!includeOverlaps) {
        return [];
      }
      return handleOverlaps(arc1, arc2);
    }
  }

  const centersVector = normalize(subtract(arc2.center, arc1.center));
  // The circles are tangent to each other
  const isOutsideTangent = centersDistance > radiusSum - epsilon;
  if (
    // circles are outside each other
    isOutsideTangent ||
    // circles are inside each other
    Math.abs(centersDistance - radiusDifference) < epsilon
  ) {
    const orientation = isOutsideTangent || arc1.radius > arc2.radius ? 1 : -1;
    const intersectionPoint = add(
      arc1.center,
      scalarMultiply(centersVector, orientation * arc1.radius),
    );

    if (
      arc1.isOnSegment(intersectionPoint) &&
      arc2.isOnSegment(intersectionPoint)
    ) {
      return [intersectionPoint];
    } else {
      return [];
    }
  }

  // The circles cross each other
  const radiusToChord =
    (arc1.radius * arc1.radius) / (2 * centersDistance) -
    (arc2.radius * arc2.radius) / (2 * centersDistance) +
    centersDistance / 2;

  const midPoint = add(
    arc1.center,
    scalarMultiply(centersVector, radiusToChord),
  );

  const halfChord = Math.sqrt(
    arc1.radius * arc1.radius - radiusToChord * radiusToChord,
  );

  const chordVector = perpendicular(centersVector);

  const p1 = add(midPoint, scalarMultiply(chordVector, halfChord));
  const p2 = add(midPoint, scalarMultiply(chordVector, -halfChord));

  const intersections = [];
  if (arc1.isOnSegment(p1) && arc2.isOnSegment(p1)) {
    intersections.push(p1);
  }
  if (arc1.isOnSegment(p2) && arc2.isOnSegment(p2)) {
    intersections.push(p2);
  }

  return intersections;
}
