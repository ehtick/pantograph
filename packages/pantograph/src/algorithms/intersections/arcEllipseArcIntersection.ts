import type { EllipseArc } from "../../models/segments/EllipseArc";
import type { Arc } from "../../models/segments/Arc";
import { ellipseEllipseIntersection } from "./ellipseEllipseIntersection";

export function arcEllipseArcIntersection(arc: Arc, ellipseArc: EllipseArc) {
  const points = ellipseEllipseIntersection(arc, ellipseArc);
  return points.filter((p) => arc.isOnSegment(p) && ellipseArc.isOnSegment(p));
}
