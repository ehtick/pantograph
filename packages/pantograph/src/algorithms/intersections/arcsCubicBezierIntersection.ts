import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Arc } from "../../models/segments/Arc.js";
import { Vector } from "../../definitions.js";
import removeDuplicateValues from "../../utils/removeDuplicateValues.js";
import { solveGenericPolynomial } from "../solvers/solvePolynomials.js";
import { CubicBezier } from "../../models/segments/CubicBezier.js";

/*
 * We use wolfram alpha to do the analitical resolution of the intersection
 * Collect[ Eliminate[{a*x^2+b*x*y+c*y^2+d * x + e * y + f == 0, p3 * t^3 + p2 * t^2  + p1 * t +p0 == x,q3*t^3+q2*t^2+q1*t+q0==y}, {x,y}], t ]
 * */

const intersectionsPolynomial = (arc: EllipseArc | Arc, curve: CubicBezier) => {
  const [[p0, p1, p2, p3], [q0, q1, q2, q3]] = curve.polynomialCoefficients;

  const el1 = arc.coefficients;
  const a = el1.x2;
  const b = el1.xy;
  const c = el1.y2;
  const d = el1.x;
  const e = el1.y;
  const f = el1.c;

  const p02 = p0 * p0;
  const p12 = p1 * p1;
  const p22 = p2 * p2;
  const p32 = p3 * p3;
  const q02 = q0 * q0;
  const q12 = q1 * q1;
  const q22 = q2 * q2;
  const q32 = q3 * q3;

  const z0 = f + d * p0 + a * p02 + e * q0 + b * p0 * q0 + c * q02;
  const z1 =
    d * p1 +
    2 * a * p0 * p1 +
    b * p1 * q0 +
    e * q1 +
    b * p0 * q1 +
    2 * c * q0 * q1;
  const z2 =
    a * p12 +
    d * p2 +
    2 * a * p0 * p2 +
    b * p2 * q0 +
    b * p1 * q1 +
    c * q12 +
    e * q2 +
    b * p0 * q2 +
    2 * c * q0 * q2;
  const z3 =
    2 * a * p1 * p2 +
    d * p3 +
    2 * a * p0 * p3 +
    b * p3 * q0 +
    b * p2 * q1 +
    b * p1 * q2 +
    2 * c * q1 * q2 +
    e * q3 +
    b * p0 * q3 +
    2 * c * q0 * q3;
  const z4 =
    a * p22 +
    2 * a * p1 * p3 +
    b * p3 * q1 +
    b * p2 * q2 +
    c * q22 +
    b * p1 * q3 +
    2 * c * q1 * q3;
  const z5 = 2 * a * p2 * p3 + b * p3 * q2 + b * p2 * q3 + 2 * c * q2 * q3;
  const z6 = a * p32 + b * p3 * q3 + c * q32;

  return [z0, z1, z2, z3, z4, z5, z6];
};

export function arcsCubicBezierIntersection(
  arc: EllipseArc | Arc,
  curve: CubicBezier
): Vector[] {
  const epsilon = Math.max(arc.precision, curve.precision);

  const polynomial = intersectionsPolynomial(arc, curve);
  const solutions = solveGenericPolynomial(polynomial, epsilon).filter((t) => {
    return t >= -curve.precision && t <= 1 + curve.precision;
  });

  return removeDuplicateValues(solutions, epsilon)
    .map((t) => {
      return curve.paramPoint(t);
    })
    .filter((p) => {
      return arc.isOnSegment(p);
    });
}
