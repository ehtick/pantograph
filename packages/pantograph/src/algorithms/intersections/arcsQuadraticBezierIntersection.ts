import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Arc } from "../../models/segments/Arc.js";
import { Vector } from "../../definitions.js";
import removeDuplicateValues from "../../utils/removeDuplicateValues.js";
import { solveQuartic } from "../solvers/solvePolynomials.js";
import { QuadraticBezier } from "../../models/segments/QuadraticBezier.js";

/*
 * We use wolfram alpha to do the analitical resolution of the intersection
 * Collect[ Eliminate[{a*x^2+b*x*y+c*y^2+d * x + e * y + f == 0, p2 * t^2  + p1 * t +p0 == x,q2*t^2+q1*t+q0==y}, {x,y}], t ]
 * */

const intersectionsPolynomial = (
  arc: EllipseArc | Arc,
  curve: QuadraticBezier
): [number, number, number, number, number] => {
  const [[p0, p1, p2], [q0, q1, q2]] = curve.polynomialCoefficients;

  const el1 = arc.coefficients;
  const a = el1.x2;
  const b = el1.xy;
  const c = el1.y2;
  const d = el1.x;
  const e = el1.y;
  const f = el1.c;

  const p02 = p0 * p0;
  const p12 = p1 * p1;
  const p22 = p2 * p2;
  const q02 = q0 * q0;
  const q12 = q1 * q1;
  const q22 = q2 * q2;

  const z0 = a * p02 + b * p0 * q0 + c * q02 + d * p0 + e * q0 + f;
  const z1 =
    2 * a * p0 * p1 +
    b * p0 * q1 +
    b * p1 * q0 +
    2 * c * q0 * q1 +
    d * p1 +
    e * q1;
  const z2 =
    2 * a * p0 * p2 +
    a * p12 +
    b * p0 * q2 +
    b * p1 * q1 +
    b * p2 * q0 +
    2 * c * q0 * q2 +
    c * q12 +
    d * p2 +
    e * q2;

  const z3 = 2 * a * p1 * p2 + b * p1 * q2 + b * p2 * q1 + 2 * c * q1 * q2;
  const z4 = a * p22 + b * p2 * q2 + c * q22;

  return [z0, z1, z2, z3, z4];
};

export function arcsQuadraticBezierIntersection(
  arc: EllipseArc | Arc,
  curve: QuadraticBezier
): Vector[] {
  const epsilon = Math.max(arc.precision, curve.precision);

  const polynomial = intersectionsPolynomial(arc, curve);
  const solutions = solveQuartic(...polynomial).filter((t) => {
    return t >= -curve.precision && t <= 1 + curve.precision;
  });

  return removeDuplicateValues(solutions, epsilon)
    .map((t) => {
      return curve.paramPoint(t);
    })
    .filter((p) => {
      return arc.isOnSegment(p);
    });
}
