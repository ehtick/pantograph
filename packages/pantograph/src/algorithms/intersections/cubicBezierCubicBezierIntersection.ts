import { Vector } from "../../definitions.js";
import { CubicBezier } from "../../models/segments/CubicBezier.js";
import { sameVector } from "../../vectorOperations.js";
import { bezierClip } from "./bezierClip.js";

export function handleOverlaps(curve1: CubicBezier, curve2: CubicBezier) {
  const commonPoints: Vector[] = [];
  const toTest: [Vector, CubicBezier][] = [
    [curve1.firstPoint, curve2],
    [curve1.lastPoint, curve2],
    [curve2.firstPoint, curve1],
    [curve2.lastPoint, curve1],
  ];

  // This could be optimised by doing only one vector => parameter conversion

  toTest.forEach(([point, curve]) => {
    if (curve.isOnSegment(point)) {
      commonPoints.push(point);
    }
  });

  if (commonPoints.length < 2) {
    // no overlap
    return null;
  }

  if (commonPoints.length === 2) {
    return [curve1.splitAt(commonPoints)[1]];
  }

  if (commonPoints.length === 3) {
    // there is one curve that is within the other
    if (
      sameVector(commonPoints[0], curve1.firstPoint) &&
      sameVector(commonPoints[1], curve1.lastPoint)
    ) {
      return [curve1];
    }
    return [curve2];
  }

  if (commonPoints.length === 4) {
    return [curve1];
  }
}

export function cubicBezierCubicBezierIntersection(
  curve1: CubicBezier,
  curve2: CubicBezier,
  includeOverlaps = false
) {
  const epsilon = Math.max(curve1.precision, curve2.precision);

  if (includeOverlaps) {
    const overlappingCurve = handleOverlaps(curve1, curve2);
    if (overlappingCurve) {
      return overlappingCurve;
    }
  }

  return bezierClip(curve1, curve2, epsilon);
}
