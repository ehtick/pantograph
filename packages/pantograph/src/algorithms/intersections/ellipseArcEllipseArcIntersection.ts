import { Vector } from "../../definitions.js";
import { EllipseArc } from "../../models/segments/EllipseArc.js";
import removeDuplicatePoints from "../../utils/removeDuplicatePoints.js";
import { sameVector } from "../../vectorOperations.js";
import { ellipseEllipseIntersection } from "./ellipseEllipseIntersection.js";

const complementArc = (arc: EllipseArc): EllipseArc => {
  const {
    firstPoint,
    lastPoint,
    center,
    majorRadius,
    minorRadius,
    tiltAngle,
    clockwise,
  } = arc;
  return new EllipseArc(
    lastPoint,
    firstPoint,
    center,
    majorRadius,
    minorRadius,
    tiltAngle,
    clockwise,
    {
      ignoreChecks: true,
      angleUnits: "rad",
    }
  );
};

const handleOverlaps = (arc1: EllipseArc, arc2: EllipseArc): EllipseArc[] => {
  // handle the case with common points at start or end first
  if (arc1.isSame(arc2)) {
    return [arc1];
  }
  // two arcs with common points at start and end (but not the same arc)

  const partialArc = (firstPoint: Vector, lastPoint: Vector) =>
    new EllipseArc(
      firstPoint,
      lastPoint,
      arc1.center,
      arc1.majorRadius,
      arc1.minorRadius,
      arc1.tiltAngle,
      arc1.clockwise,
      { ignoreChecks: true, angleUnits: "rad" }
    );

  const points = removeDuplicatePoints(
    [
      arc2.isOnSegment(arc1.firstPoint) ? arc1.firstPoint : null,
      arc2.isOnSegment(arc1.lastPoint) ? arc1.lastPoint : null,
      arc1.isOnSegment(arc2.firstPoint) ? arc2.firstPoint : null,
      arc1.isOnSegment(arc2.lastPoint) ? arc2.lastPoint : null,
    ].filter((p) => p !== null) as Vector[]
    // We sort by the param value of the first arc. This means that the points
    // will be sorted with the same orientation than arc1
  ).sort((a, b) => arc1.pointToParam(a) - arc1.pointToParam(b));

  if (points.length === 0) return [];
  // We consider the case when the arcs touch only on
  // the last point. We consider that they do not overlap there
  //
  // We might want to revisit this choice
  else if (points.length === 1) return [];
  else if (points.length === 2) {
    // Similar to the case with length 1, we ignore the double overlapping point
    if (arc1.isSame(complementArc(arc2))) return [];
    return [partialArc(points[0], points[1])];
  } else if (points.length === 3) {
    // Similar to the case with length 1, we ignore the single overlapping point
    const startIndex =
      sameVector(points[0], arc2.lastPoint) ||
      sameVector(points[0], arc2.firstPoint)
        ? 1
        : 0;
    return [partialArc(points[0 + startIndex], points[1 + startIndex])];
  } else if (points.length === 4) {
    return [partialArc(points[0], points[1]), partialArc(points[2], points[3])];
  }
  throw new Error("Bug in the ellipse arc ellipse arc overlap algorithm");
};

export function ellipseArcEllipseArcIntersection(
  arc1: EllipseArc,
  arc2: EllipseArc,
  includeOverlaps = false
): Vector[] | EllipseArc[] {
  const epsilon = Math.max(arc1.precision, arc2.precision);
  const sameEllipse =
    sameVector(arc1.center, arc2.center) &&
    Math.abs(arc1.majorRadius - arc2.majorRadius) < epsilon &&
    Math.abs(arc1.minorRadius - arc2.minorRadius) < epsilon &&
    (Math.abs(arc1.tiltAngle - arc2.tiltAngle) < epsilon ||
      Math.abs(Math.abs(arc1.tiltAngle - arc2.tiltAngle) - Math.PI) < epsilon);
  if (sameEllipse) {
    if (includeOverlaps) return handleOverlaps(arc1, arc2);
    else return [];
  }
  const points = ellipseEllipseIntersection(arc1, arc2);
  return points.filter((p) => arc1.isOnSegment(p) && arc2.isOnSegment(p));
}
