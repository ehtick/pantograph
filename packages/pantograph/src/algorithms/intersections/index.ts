import { Line } from "../../models/segments/Line";

import type { Vector } from "../../definitions";
import { Segment } from "../../models/segments/Segment";
import { Arc } from "../../models/segments/Arc";
import { lineArcIntersection } from "./lineArcIntersection";
import { lineLineIntersection } from "./lineLineIntersection";
import { arcArcIntersection } from "./arcArcIntersection";

export function findIntersections(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): Vector[] {
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      false,
      precision
    );
    if (intersection === null) return [];
    return [intersection as Vector];
  }
  if (segment1 instanceof Line && segment2 instanceof Arc) {
    return lineArcIntersection(segment1, segment2, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Line) {
    return lineArcIntersection(segment2, segment1, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    return arcArcIntersection(segment1, segment2, false, precision) as Vector[];
  }

  throw new Error("Not implemented");
}

export function findIntersectionsAndOverlaps(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): { intersections: Vector[]; overlaps: Segment[]; count: number } {
  // If we have two lines, checks are fast enough to not use bounding boxes
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (intersection === null)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersection instanceof Line)
      return { intersections: [], overlaps: [intersection], count: 1 };
    return { intersections: [intersection], overlaps: [], count: 1 };
  }

  if (!segment1.boundingBox.overlaps(segment2.boundingBox)) {
    return { intersections: [], overlaps: [], count: 0 };
  }

  if (segment1 instanceof Line && segment2 instanceof Arc) {
    const intersections = lineArcIntersection(segment1, segment2, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Line) {
    const intersections = lineArcIntersection(segment2, segment1, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    const intersections = arcArcIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (!intersections.length)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersections[0] instanceof Arc)
      return {
        intersections: [],
        overlaps: intersections as Arc[],
        count: intersections.length,
      };
    return {
      intersections: intersections as Vector[],
      overlaps: [],
      count: intersections.length,
    };
  }

  throw new Error("Not implemented");
}
