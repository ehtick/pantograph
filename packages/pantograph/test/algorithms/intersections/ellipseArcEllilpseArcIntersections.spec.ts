import { describe, it, expect } from "vitest";

import { ellipseArcEllipseArcIntersection } from "../../../src/algorithms/intersections/ellipseArcEllipseArcIntersection";
import { EllipseArc } from "../../../src/models/segments/EllipseArc";

//import { debugImg, dpnt } from "../../debug";

describe("arcEllipseArcIntersection", () => {
  describe("should return the intersection points of an arc and an ellipse arc", () => {
    it("should work for one direction", () => {
      const arc = new EllipseArc(
        [0, -0.5],
        [3, 0],
        [0, 0],
        3,
        0.5,
        0,
        true
      ).rotate(45);
      const ellipseArc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false)
        .translateX(-2)
        .rotate(5);
      const intersections = ellipseArcEllipseArcIntersection(arc, ellipseArc);

      //debugImg([arc, ellipseArc, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(3);
      expect(intersections).toMatchSnapshot();
    });

    it("should work for the opposite direction", () => {
      const arc = new EllipseArc(
        [0, -0.5],
        [3, 0],
        [0, 0],
        3,
        0.5,
        0,
        false
      ).rotate(45);
      const ellipseArc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false)
        .translateX(-2)
        .rotate(5);
      const intersections = ellipseArcEllipseArcIntersection(arc, ellipseArc);

      //debugImg([arc, ellipseArc, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(1);
      expect(intersections).toMatchSnapshot();
    });
  });

  it("should work for two common points", () => {
    const arc = new EllipseArc([0, -0.5], [3, 0], [0, 0], 3, 0.5, 0, true);
    const ellipseArc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false)
      .translateX(-2)
      .rotate(5)
      .translateY(1);

    const intersections = ellipseArcEllipseArcIntersection(arc, ellipseArc);

    //debugImg([arc, ellipseArc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(2);
    expect(intersections).toMatchSnapshot();
  });

  it("should work for tangent common points", () => {
    const arc = new EllipseArc([0, -1], [3, 0], [0, 0], 3, 1, 0, true);
    const ellipseArc = new EllipseArc(
      [0, 0],
      [2, 1],
      [2, 0],
      2,
      1,
      0,
      false
    ).translateX(-2);

    const intersections = ellipseArcEllipseArcIntersection(arc, ellipseArc);

    //debugImg([arc, ellipseArc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(2);
    expect(intersections).toMatchSnapshot();
  });

  it("should work for no intersection", () => {
    const arc = new EllipseArc([0, -0.5], [3, 0], [0, 0], 3, 0.5, 0, true)
      .mirror("x")
      .translateY(3);
    const ellipseArc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, false)
      .translateX(-2)
      .rotate(1);

    const intersections = ellipseArcEllipseArcIntersection(arc, ellipseArc);

    //debugImg([arc, ellipseArc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(0);
  });

  it("should work for axis aligned ellipses", () => {
    const arc = new EllipseArc([0, -0.5], [3, 0], [0, 0], 3, 0.5, 0, true);
    const ellipseArc = new EllipseArc(
      [0, 0],
      [2, 1],
      [2, 0],
      2,
      1,
      0,
      false
    ).translateX(-2);
    const intersections = ellipseArcEllipseArcIntersection(
      arc,
      ellipseArc,
      true
    );

    //debugImg([arc, ellipseArc, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(2);
    expect(intersections).toMatchSnapshot();
  });

  it("works with natural ellipses", () => {
    const arc1 = new EllipseArc([-8, 5], [-10, 0], [-8, 0], 5, 2, 90);
    const arc2 = new EllipseArc(
      [-6.123724356957945, 3.5355339059327378],
      [-24.406964702329226, -5.50455030127395],
      [-5.088448176547862, -0.3281693992235354],
      20,
      4,
      15
    );
    const intersections = ellipseArcEllipseArcIntersection(arc1, arc2, true);

    //debugImg([arc1, arc2, ...intersections.map((p) => dpnt(p))]);
    expect(intersections.length).toBe(1);
    expect(intersections).toMatchSnapshot();
  });

  describe("overlapping ellipses", () => {
    it("works with one overlap", () => {
      const ellipseArc = new EllipseArc(
        [0, 0],
        [2, 1],
        [2, 0],
        2,
        1,
        0,
        false
      ).translateX(-2);

      const p0 = ellipseArc.paramPoint(-0.2);
      const p1 = ellipseArc.paramPoint(0.3);

      const arc = new EllipseArc(p0, p1, [0, 0], 2, 1, 0, false);

      const intersections = ellipseArcEllipseArcIntersection(
        arc,
        ellipseArc,
        true
      );

      //debugImg([arc, ellipseArc, ...intersections]);
      expect(intersections.length).toBe(1);
      expect(intersections[0]).toBeInstanceOf(EllipseArc);
      expect(intersections).toMatchSnapshot();
    });

    it("works with two overlaps", () => {
      const ellipseArc = new EllipseArc(
        [0, 0],
        [2, 1],
        [2, 0],
        2,
        1,
        0,
        false
      ).translateX(-2);

      const p0 = ellipseArc.paramPoint(0.4);
      const p1 = ellipseArc.paramPoint(0.3);

      const arc = new EllipseArc(p0, p1, [0, 0], 2, 1, 0, false);

      const intersections = ellipseArcEllipseArcIntersection(
        arc,
        ellipseArc,
        true
      );

      //debugImg([arc, ellipseArc, ...intersections]);
      expect(intersections.length).toBe(2);
      expect(intersections[0]).toBeInstanceOf(EllipseArc);
      expect(intersections).toMatchSnapshot();
    });

    it("works with two overlaps, with rotated ellipses", () => {
      const ellipseArc = new EllipseArc(
        [0, 0],
        [2, 1],
        [2, 0],
        2,
        1,
        0,
        false
      ).translateX(-2);

      const p0 = ellipseArc.paramPoint(0.4);
      const p1 = ellipseArc.paramPoint(0.3);

      const arc = new EllipseArc(p0, p1, [0, 0], 2, 1, 0, false);

      const intersections = ellipseArcEllipseArcIntersection(
        arc.rotate(33),
        ellipseArc.rotate(33),
        true
      );

      //debugImg([arc.rotate(33), ellipseArc.rotate(33), ...intersections]);
      expect(intersections.length).toBe(2);
      expect(intersections[0]).toBeInstanceOf(EllipseArc);
      expect(intersections).toMatchSnapshot();
    });

    it("works with no overlaps", () => {
      const ellipseArc = new EllipseArc(
        [0, 0],
        [2, 1],
        [2, 0],
        2,
        1,
        0,
        false
      ).translateX(-2);

      const p0 = ellipseArc.paramPoint(-0.2);
      const p1 = ellipseArc.paramPoint(-0.1);

      const arc = new EllipseArc(p0, p1, [0, 0], 2, 1, 0, false);

      const intersections = ellipseArcEllipseArcIntersection(
        arc,
        ellipseArc,
        true
      );

      //debugImg([arc, ellipseArc, ...intersections]);
      expect(intersections.length).toBe(0);
    });
  });
});
