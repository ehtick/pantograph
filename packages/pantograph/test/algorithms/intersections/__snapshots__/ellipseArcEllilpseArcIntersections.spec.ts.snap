// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`arcEllipseArcIntersection > overlapping ellipses > works with one overlap 1`] = `
[
  EllipseArc {
    "_coefficients": null,
    "center": [
      0,
      0,
    ],
    "clockwise": false,
    "firstPoint": [
      -2,
      0,
    ],
    "lastPoint": [
      -0.31286893008046207,
      -0.9876883405951377,
    ],
    "majorRadius": 2,
    "minorRadius": 1,
    "precision": 0.000001,
    "segmentType": "ELLIPSE_ARC",
    "tiltAngle": 0,
  },
]
`;

exports[`arcEllipseArcIntersection > overlapping ellipses > works with two overlaps 1`] = `
[
  EllipseArc {
    "_coefficients": null,
    "center": [
      0,
      0,
    ],
    "clockwise": false,
    "firstPoint": [
      0.6180339887498945,
      -0.9510565162951536,
    ],
    "lastPoint": [
      0,
      1,
    ],
    "majorRadius": 2,
    "minorRadius": 1,
    "precision": 0.000001,
    "segmentType": "ELLIPSE_ARC",
    "tiltAngle": 0,
  },
  EllipseArc {
    "_coefficients": null,
    "center": [
      0,
      0,
    ],
    "clockwise": false,
    "firstPoint": [
      -2,
      0,
    ],
    "lastPoint": [
      -0.31286893008046207,
      -0.9876883405951377,
    ],
    "majorRadius": 2,
    "minorRadius": 1,
    "precision": 0.000001,
    "segmentType": "ELLIPSE_ARC",
    "tiltAngle": 0,
  },
]
`;

exports[`arcEllipseArcIntersection > overlapping ellipses > works with two overlaps, with rotated ellipses 1`] = `
[
  EllipseArc {
    "_coefficients": null,
    "center": [
      0,
      0,
    ],
    "clockwise": false,
    "firstPoint": [
      1.0363094196341955,
      -0.4610176734302223,
    ],
    "lastPoint": [
      -0.5446390350150271,
      0.838670567945424,
    ],
    "majorRadius": 2,
    "minorRadius": 1,
    "precision": 0.000001,
    "segmentType": "ELLIPSE_ARC",
    "tiltAngle": 0.5759586531581287,
  },
  EllipseArc {
    "_coefficients": null,
    "center": [
      0,
      0,
    ],
    "clockwise": false,
    "firstPoint": [
      -1.677341135890848,
      -1.0892780700300542,
    ],
    "lastPoint": [
      0.2755396614342709,
      -0.9987457737252043,
    ],
    "majorRadius": 2,
    "minorRadius": 1,
    "precision": 0.000001,
    "segmentType": "ELLIPSE_ARC",
    "tiltAngle": 0.5759586531581287,
  },
]
`;

exports[`arcEllipseArcIntersection > should return the intersection points of an arc and an ellipse arc > should work for one direction 1`] = `
[
  [
    -0.3401205058312131,
    -1.0104433405109907,
  ],
  [
    -1.4066529619990422,
    -0.8030218877085064,
  ],
  [
    0.3401205058312048,
    1.0104433405109925,
  ],
]
`;

exports[`arcEllipseArcIntersection > should return the intersection points of an arc and an ellipse arc > should work for the opposite direction 1`] = `
[
  [
    1.4066529619990515,
    0.8030218877085047,
  ],
]
`;

exports[`arcEllipseArcIntersection > should work for axis aligned ellipses 1`] = `
[
  [
    -1.8371173883849605,
    -0.395284697025575,
  ],
  [
    1.8371172820973773,
    0.3952847107472433,
  ],
]
`;

exports[`arcEllipseArcIntersection > should work for tangent common points 1`] = `
[
  [
    -8.940696716308594e-8,
    -0.9999999999999996,
  ],
  [
    0.00038102339293144435,
    0.9999999919345096,
  ],
]
`;

exports[`arcEllipseArcIntersection > should work for two common points 1`] = `
[
  [
    -1.7517091212757456,
    0.4059117020306511,
  ],
  [
    1.4940750915928842,
    0.43358132713599973,
  ],
]
`;
